File: Python3/aoc2020/solutions/day_02.py
import re
from typing import Callable, Iterable

from .solution import Solution

PASSWORD_REGEX_PATTERN = re.compile(r"^(\d+)-(\d+) (\w): (\w+)$")


def current_policy(first_number: int, second_number: int, letter: str, password: str) -> bool:
    return (password[first_number - 1] == letter) ^ (password[second_number - 1] == letter)


def old_policy(first_number: int, second_number: int, letter: str, password: str) -> bool:
    occurences_of_letter = sum(ch == letter for ch in password)

    return first_number <= occurences_of_letter <= second_number


def is_password_valid(password: str, password_policy: Callable[[int, int, str, str], bool]) -> bool:
    password_text_match = PASSWORD_REGEX_PATTERN.fullmatch(password)

    if password_text_match is None:
        return False

    first_number_str, second_number_str, letter, password_text = password_text_match.groups()
    first_number = int(first_number_str, base=10)
    second_number = int(second_number_str, base=10)

    return password_policy(first_number, second_number, letter, password_text)


def count_old_valid_passwords(passwords: Iterable[str]) -> int:
    return sum(is_password_valid(password, old_policy) for password in passwords)


def count_current_valid_passwords(passwords: Iterable[str]) -> int:
    return sum(is_password_valid(password, current_policy) for password in passwords)


class Day02(Solution):

    def first_task(self, passwords_text: str) -> str:
        passwords = [password.strip()
                     for password in passwords_text.splitlines()]

        return str(count_old_valid_passwords(passwords))

    def second_task(self, passwords_text: str) -> str:
        passwords = [password.strip()
                     for password in passwords_text.splitlines()]

        return str(count_current_valid_passwords(passwords))

File: Rust/src/solutions/day_02.rs
use super::solution::{Error as ChallengeErr, Solution};
use lazy_static::lazy_static;
use regex::Regex;

lazy_static! {
    static ref PASSWORD_PATTERN: Regex = Regex::new(r"^(\d+)-(\d+) (\w): (\w+)$").unwrap();
}

struct PasswordDefinition<'a> {
    first_number: i32,
    second_number: i32,
    letter: char,
    password: &'a str,
}

impl PasswordDefinition<'_> {
    fn from(password_text: &str) -> PasswordDefinition {
        let captures = PASSWORD_PATTERN.captures(password_text).unwrap();

        let first_number: i32 = captures.get(1).unwrap().as_str().parse().unwrap();
        let second_number: i32 = captures.get(2).unwrap().as_str().parse().unwrap();
        let letter: char = captures.get(3).unwrap().as_str().chars().next().unwrap();
        let password: &str = captures.get(4).unwrap().as_str();

        PasswordDefinition {
            first_number,
            second_number,
            letter,
            password,
        }
    }

    fn align_with_old_policy(&self) -> bool {
        let occurences_of_letter = self
            .password
            .chars()
            .filter(|&ch| ch == self.letter)
            .count() as i32;

        self.first_number <= occurences_of_letter && occurences_of_letter <= self.second_number
    }

    fn align_with_current_policy(&self) -> bool {
        (self
            .password
            .chars()
            .nth((self.first_number - 1) as usize)
            .unwrap()
            == self.letter)
            ^ (self
                .password
                .chars()
                .nth((self.second_number - 1) as usize)
                .unwrap()
                == self.letter)
    }
}

pub struct Day02 {}

impl Day02 {
    fn count_old_valid_passwords(passwords: &[&str]) -> i32 {
        return passwords
            .iter()
            .map(|password| PasswordDefinition::from(password))
            .filter(|def| def.align_with_old_policy())
            .count() as i32;
    }

    fn count_current_valid_passwords(passwords: &[&str]) -> i32 {
        return passwords
            .iter()
            .map(|password| PasswordDefinition::from(password))
            .filter(|def| def.align_with_current_policy())
            .count() as i32;
    }
}

impl Solution for Day02 {
    fn first_task(&self, input: &str) -> Result<String, ChallengeErr> {
        let passwords_list: Vec<&str> = input.lines().collect();

        Ok(Day02::count_old_valid_passwords(&passwords_list).to_string())
    }

    fn second_task(&self, input: &str) -> Result<String, ChallengeErr> {
        let passwords_list: Vec<&str> = input.lines().collect();

        Ok(Day02::count_current_valid_passwords(&passwords_list).to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example_count_old_valid_passwords() {
        let test_passwords = ["1-3 a: abcde", "1-3 b: cdefg", "2-9 c: ccccccccc"];

        assert_eq!(Day02::count_old_valid_passwords(&test_passwords), 2);
    }

    #[test]
    fn example_count_current_valid_passwords() {
        let test_passwords = ["1-3 a: abcde", "1-3 b: cdefg", "2-9 c: ccccccccc"];

        assert_eq!(Day02::count_current_valid_passwords(&test_passwords), 1);
    }
}
